File: apps/frontend/src/api/scanner.ts
import { apiFetch } from "./client";

// P8 §2/§3 — Discovery scanner. A saved boolean criterion (over supported
// indicator names) + a universe spec; runs record matched/skipped symbols.

export type UniverseKind = "discovery_feeds" | "watchlist" | "symbols";

export interface UniverseSpec {
  kind: UniverseKind;
  symbols?: string[] | null;
}

export interface ScannerDefinition {
  id: number;
  name: string;
  criteria: string;
  universe_kind: UniverseKind;
  universe_symbols: string[] | null;
  timeframe: string;
  created_at: string;
  updated_at: string;
}

export interface ScannerDefinitionInput {
  name: string;
  criteria: string;
  universe: UniverseSpec;
  timeframe?: string;
}

export interface ScannerMatchItem {
  symbol: string;
  values: Record<string, number>;
}

export interface ScannerSkipItem {
  symbol: string;
  reason: string;
}

export interface ScannerRunSummary {
  id: number;
  scanner_definition_id: number;
  run_at: string;
  status: string;
  universe_size: number;
  evaluated_count: number;
  matched_count: number;
  skipped_count: number;
  error: string | null;
}

export interface ScannerRun extends ScannerRunSummary {
  criteria_snapshot: string;
  universe_kind: string;
  timeframe: string;
  matched: ScannerMatchItem[];
  skipped: ScannerSkipItem[];
}

export interface ScannerVocabulary {
  indicators: string[];
  fields: string[];
}

export const scannerApi = {
  vocabulary: () => apiFetch<ScannerVocabulary>("/api/v1/scanner/vocabulary"),
  list: () => apiFetch<ScannerDefinition[]>("/api/v1/scanner/definitions"),
  create: (body: ScannerDefinitionInput) =>
    apiFetch<ScannerDefinition>("/api/v1/scanner/definitions", {
      method: "POST",
      body: JSON.stringify(body),
    }),
  update: (id: number, body: ScannerDefinitionInput) =>
    apiFetch<ScannerDefinition>(`/api/v1/scanner/definitions/${id}`, {
      method: "PUT",
      body: JSON.stringify(body),
    }),
  remove: (id: number) =>
    apiFetch<void>(`/api/v1/scanner/definitions/${id}`, { method: "DELETE" }),
  run: (id: number) =>
    apiFetch<ScannerRun>(`/api/v1/scanner/definitions/${id}/run`, {
      method: "POST",
    }),
  listRuns: (id: number) =>
    apiFetch<ScannerRunSummary[]>(`/api/v1/scanner/definitions/${id}/runs`),
  getRun: (runId: number) =>
    apiFetch<ScannerRun>(`/api/v1/scanner/runs/${runId}`),
};

File: apps/backend/app/api/v1/schemas/scanner.py
"""Pydantic models for the Discovery scanner endpoints (P8 §2)."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class UniverseSpec(BaseModel):
    kind: str  # discovery_feeds | watchlist | symbols
    symbols: list[str] | None = None  # required for kind="symbols"


class ScannerDefinitionCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    criteria: str = Field(min_length=1)
    universe: UniverseSpec
    timeframe: str = "1Day"


class ScannerDefinitionResponse(BaseModel):
    id: int
    name: str
    criteria: str
    universe_kind: str
    universe_symbols: list[str] | None
    timeframe: str
    created_at: datetime
    updated_at: datetime


class ScannerVocabulary(BaseModel):
    indicators: list[str]  # supported bare indicator names (drift-proof)
    fields: list[str]  # bar fields (open/high/low/close/volume/price)


class ScannerMatchItem(BaseModel):
    symbol: str
    values: dict[str, float]


class ScannerSkipItem(BaseModel):
    symbol: str
    reason: str


class ScannerRunSummary(BaseModel):
    id: int
    scanner_definition_id: int
    run_at: datetime
    status: str
    universe_size: int
    evaluated_count: int
    matched_count: int
    skipped_count: int
    error: str | None


class ScannerRunResponse(ScannerRunSummary):
    criteria_snapshot: str
    universe_kind: str
    timeframe: str
    matched: list[ScannerMatchItem]
    skipped: list[ScannerSkipItem]
